theano/gpuarray/GpuMaxAndArgmax.py

```diff
@@ -0,0 +1,209 @@
+        else:
+            axes = tuple(axes)
+            # axes = tuple(int(ax) for ax in axes)
+        max[0] = theano._asarray(numpy.max(X, axes), dtype=node.outputs[0].dtype)
+        # Numpy does not support multiple axes for argmax
+        # Work around
+        keep_axes = numpy.array([i for i in range(X.ndim) if i not in axes], dtype='int64')
+        # Not-reduced axes in front
+        transposed_x = numpy.transpose(X, numpy.concatenate((keep_axes, axes)))
+        kept_shape = transposed_x.shape[:len(keep_axes)]
+        reduced_shape = transposed_x.shape[len(keep_axes):]
+        new_shape = kept_shape + (numpy.prod(reduced_shape),)
+        reshaped_x = transposed_x.reshape(new_shape)
+
+        max_idx[0] = theano._asarray(numpy.argmax(reshaped_x, axis=-1), dtype='int64')
+
+    def c_headers(self):
+        return ['<numpy_compat.h>', '<gpuarray_helper.h>']
+
+    def c_header_dirs(self):
+        return [pygpu.get_include(), os.path.dirname(__file__)]
+
+    def c_code(self, node, name, input_names, output_names, sub):
+        # Recall: X, axes = input_names
+        # Recall: max, argmax = output_names
+        # Recall: fail = sub['fail']
+        max_typecode = pygpu.gpuarray.dtype_to_typecode(node.inputs[0].dtype)
+        argmax_typecode = pygpu.gpuarray.dtype_to_typecode(self.argmax_dtype)
+        # axes_ctype = pygpu.gpuarray.dtype_to_ctype(node.inputs[1].dtype)
+        axes_ctype = 'int64_t'
+        ret = """
+        GpuArray* %(name)s_input = &%(X)s->ga;
+        size_t %(name)s_input_ndim = PyGpuArray_NDIM(%(X)s);
+        """
+        if NoneConst.equals(node.inputs[1]):
+            ret += """
+            unsigned  %(name)s_redux_len = %(name)s_input_ndim;
+            unsigned* %(name)s_axes_to_reduce = new unsigned[%(name)s_redux_len];
+            for(unsigned i = 0; i < %(name)s_redux_len; ++i) {
+                %(name)s_axes_to_reduce[i] = i;
+            }
+            """
+        else:
+            assert node.inputs[1].ndim == 1
+            ret += """
+            unsigned  %(name)s_redux_len = PyArray_DIM(%(axes)s, 0);
+            unsigned* %(name)s_axes_to_reduce = new unsigned[%(name)s_redux_len];
+            for(unsigned i = 0; i < %(name)s_redux_len; ++i) {
+                %(name)s_axes_to_reduce[i] =
+                    (unsigned)( ((%(axes_ctype)s*)PyArray_DATA(%(axes)s)) [i * (PyArray_STRIDES(%(axes)s)[0] / sizeof(%(axes_ctype)s))] );
+            }
+            """
+        ret += """
+        size_t  %(name)s_output_ndim = %(name)s_input_ndim - %(name)s_redux_len;
+        size_t* %(name)s_output_dims = NULL;
+        if(%(name)s_output_ndim == 0) {
+            /* Current backend function GpuArray_maxandargmax does not work when
+             * all axes need to be reduced. So to handle this case, we create a view
+             * of the input as a matrix with 1 row and as many columns as elements
+             * in the input, so that the 2nd dimenson of the matrix will be reduced. */
```

Comment:
Yes, I have an error when all axes need to be reduced.

Indeed I think it comes from libgpuarray:
In file https://github.com/Theano/libgpuarray/blob/master/src/gpuarray_reduction.c
In function ```static int maxandargmaxCheckargs (maxandargmax_ctx* ctx)```:
There is this code:
```
	/* Insane src or reduxLen? */
	if(!ctx->dstMax || !ctx->dstArgmax || !ctx->src || ctx->src->nd == 0 ||
	    ctx->reduxLen == 0 || ctx->reduxLen >= ctx->src->nd){
		return ctx->ret=GA_INVALID_ERROR;
	}
```
I think the problem is the test ```ctx->reduxLen >= ctx->src->nd```
That should better be ```ctx->reduxLen > ctx->src->nd``` (```>``` instead of ```>=```, as the case ```ctx->reduxLen == ctx->src->nd``` is a valid case, corresponding to the reduction of all axes).